simulator/src/main/java/kcl/teamIndexZero/traffic/simulator/data/trafficLight/TrafficLightController.java

```diff
@@ -1,17 +1,118 @@
+
+       /**
+         * Constructor
+         */
+        public TrafficLightController(){
+            TrafficLightSinglesList = new ArrayList<TrafficLight>();
+            TrafficLightSetList = new ArrayList<TrafficLightInSet>();
+        }
+
+        public long formatTimeToLong(LocalDateTime date) {
+            long milliseconds = ChronoUnit.MILLIS.between(start, date);
+            return milliseconds;
+        }
+
+       /**
+         * Adds single traffic light to the List of all single traffic lights
+         *
+         * @param trafficLight object to be added to the list
+         */
+        public void addTrafficlight(TrafficLight trafficLight){
+
+            if (!(trafficLight == null)) {
+                TrafficLightSinglesList.add(trafficLight);
+                LOG.log("Added the following traffic lights to the set of all single Traffic Lights: ", trafficLight.getTrafficLightID() );
+            }
+            else {
+                LOG.log_Error("Error while adding to TrafficLightLnSet to the set");
+            }
+
+        }
+
+        /**
+         * Returns a list List of all single traffic lights
+         *
+         */
+         public List<TrafficLight> getSingleSet() {return TrafficLightSinglesList;}
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+         public void tick(SimulationTick tick) {
+            CurrentTime = formatTimeToLong(simulationTick.getSimulatedTime());
+
+            /*when single traffic lights do not work in a synchronous way with the ticks*/
+            model.TrafficLightDelay = CurrentTime - lastChange;
+            if ((model.TrafficLightDelay >= timer) && (model.currentState == TrafficLightState.GREEN)) {
+                model.currentState = TrafficLightState.RED;
+            }
+
+            /*when set's traffic lights do not work in a synchronous way with the ticks*/
+            modelSet.TrafficLightInSetDelay = CurrentTime - lastChange;
+            if ((modelSet.TrafficLightInSetDelay >= timer) && (modelSet.currentState == TrafficLightState.GREEN)){
+                modelSet.currentState = TrafficLightState.RED;
+            }
 
+            if ((CurrentTime - lastChange) > timer) {
+                if (modelSet != null) {
+                    TrafficLightsInSetRule.changeStateofSet(TrafficLightSetList);
+                    TrafficLightRule.changeStateofSingleTrafficLights(TrafficLightSinglesList);
+                    lastChange = CurrentTime;
+                }
+            }
+        }
 
-    @Override
-    public void tick(SimulationTick tick) {
+       /**
+         * Adds rule to the one traffic light
+         */
+         public void addRule(TrafficLightRule rule){
+            rule.changeStateofSingleTrafficLights(TrafficLightSinglesList);
+        }
+       /**
+         * Adds rule to the one traffic light set
+         */
+         public void addRule(TrafficLightsInSetRule rule){
```

Comment:
So how are the TrafficLightInSet linked together so that they work synchronously?